### ISSUE TITLE
Logging: Found execution with unknown task-name

### ISSUE THREAD

**Issue by ruifscouto** [kagkarlsson/db-scheduler#18]
Hello,

The use case scenario is that I have a single scheduler with 5 scheduler threads, 10s heartbeat interval and polling interval of 5s. This scheduler has to start several recurring tasks both on a FixedDelay & Daily.
I'm still not sure if the issue happens for the Daily task, but for the FixedDelay tasks I get a huge amount of logging, for each task, saying:
"internal-execute-due-pool-7-thread-1] com.github.kagkarlsson.scheduler.TaskResolver.resolve : Found execution with unknown task-name <TASK_NAME>".
It seems that task execution itself is not affected.

I'm convinced it might be related to threading, but can't be sure.
Is this an issue for anyone else? Could I be missing something?

Thanks,
Rui

**Comment by kagkarlsson** [comment]
The error you are describing is caused by the scheduler finding an execution in the `scheduled_task` table that it cannot map to a `Task`. The mapping is set up when instantiating the scheduler, as is done in this [example](https://github.com/kagkarlsson/db-scheduler#ad-hoc-tasks-1). Any tasks in the call `startTasks(...)` will automatically get a mapping (i.e. RecurringTasks).

Did you supply the scheduler with all ad-hoc tasks in the call to `.create(dataSource, task1, task2, ...)` ?

**Comment by ruifscouto** [comment]
I have exactly one ad-hoc task and that one is being passed on to the create method. 
Recurrent tasks are being "added" to the scheduler using `startTasks(...)`, before I build & start the scheduler. The method `startTasks(...)` is called multiple times: I do a set of if statements to check for profiles which indicate the tasks that should be started.

```
if(x) {
    builder.startTasks(taskA);
}

if(y) {
    builder.startTasks(taskB);
}

```

I've tried bundling a couple of tasks in a single `startTasks()` and the same issue occurs. Also having the issue if I have a simple scheduler with an ad-hoc task (so passed in the `create` method).
Could this somehow be an issue? I do know that on TaskResolver.java, taskMap field has all the tasks in some scenarios and none in others, causing the log message.

**Comment by kagkarlsson** [comment]
Calling `startTaks` multiple times should work.

Sounds to me like you have some old execution in the database for which there is no mapping (i.e. another task-name than your current ad-hoc task). Could that be the case?

**Comment by kagkarlsson** [comment]
If that is the case, you could add a no-op task with that name to "drain" it from the db.

**Comment by ruifscouto** [comment]
That is not the case. It's throwing the log even for tasks I am sure were added to startTasks (and that I'm sure are running).

**Comment by ruifscouto** [comment]
This is a sample of the execution logs:
```
2018-01-12 10:26:59 +0000, [INFO] [internal-pool-2-thread-3] TaskA.execute : TaskA start
2018-01-12 10:26:59 +0000, [INFO] [internal-pool-2-thread-3] TaskA.execute : TaskA finish
2018-01-12 10:27:04 +0000, [INFO] [internal-pool-2-thread-4] TaskB.execute : TaskB done
2018-01-12 10:27:15 +0000, [WARN] [internal-execute-due-pool-7-thread-1] com.github.kagkarlsson.scheduler.TaskResolver.resolve : Found execution with unknown task-name 'TaskA'
2018-01-12 10:27:19 +0000, [INFO] [internal-pool-2-thread-5] TaskA.execute : TaskA start
2018-01-12 10:27:19 +0000, [INFO] [internal-pool-2-thread-5] TaskA.execute : TaskA finish
2018-01-12 10:27:25 +0000, [WARN] [internal-execute-due-pool-7-thread-1] com.github.kagkarlsson.scheduler.TaskResolver.resolve : Found execution with unknown task-name 'TaskB'
2018-01-12 10:27:29 +0000, [INFO] [internal-pool-2-thread-2] TaskB.execute : TaskB done
2018-01-12 10:27:35 +0000, [WARN] [internal-execute-due-pool-7-thread-1] com.github.kagkarlsson.scheduler.TaskResolver.resolve : Found execution with unknown task-name 'TaskA'
2018-01-12 10:27:39 +0000, [INFO] [internal-pool-2-thread-1] TaskA.execute : TaskA start
2018-01-12 10:27:39 +0000, [INFO] [internal-pool-2-thread-1] TaskA.execute : TaskA finish
2018-01-12 10:27:50 +0000, [WARN] [internal-execute-due-pool-7-thread-1] com.github.kagkarlsson.scheduler.TaskResolver.resolve : Found execution with unknown task-name 'TaskB'
2018-01-12 10:27:54 +0000, [INFO] [internal-pool-2-thread-3] TaskB.execute : TaskB done
2018-01-12 10:27:55 +0000, [WARN] [internal-execute-due-pool-7-thread-1] com.github.kagkarlsson.scheduler.TaskResolver.resolve : Found execution with unknown task-name 'TaskA'
```

scheduled_tasks table contains exactly two records: one for task_name TaskA and another for task_name TaskB.

**Comment by ruifscouto** [comment]
I think I know what the problem is. I basically require two schedulers: one for recurrent tasks and one for a specific oneTimeTask that has to trigger it self to re-execute again after a given interval (this because we can't have taskData in a recurrent task).
The problem is that when tasks of the first scheduler run, somehow, they have access to the other scheduler tasks and the verification on TaskResolver.taskMap actually fails.

**Comment by kagkarlsson** [comment]
Ok, that explains things :)

All schedulers using the same database-schema should know about that same tasks, i.e. have the same tasks in their calls to `create(..)` and `startTasks(...)`. It is not yet possible to have multiple scheduler tables in the same schema, but I might make the table-name configurable in the future, especially if it is a feature-request.

But I see your problem. You want to schedule execution of TaskY at the execution-time of TaskX, but to do that you need a reference to the scheduler. I think the solution might be to extend the `ExecutionHandler` with a reference to the `Scheduler` (or rather `SchedulerClient`). Something like: 

```
public interface ExecutionHandler<T> {
	void execute(TaskInstance<T> taskInstance, ExecutionContext executionContext, SchedulerClient scheduler);
}
```

Would that work for you?

**Comment by ruifscouto** [comment]
Hello,

Yes, that would work. I would just use the Scheduler added as a parameter in the execute method to reschedule the one-time task, instead of having to create an entirely different scheduler.
Regarding the multiple schedulers issue, wouldn't it be possible to just add a new column in the table, which holds the scheduler that task is associated to? 

Also, while we're at it, is there a specific reason that recurrentTasks with task_data aren't implemented?

Thanks

**Comment by kagkarlsson** [comment]
Having multiple schedulers in the same application feels like something of a special-case, so I don't want the data-model to be affected by it. Having two different tables for the two schedulers seems like a better idea for that case. 

How quickly do you need the scheduler-parameter feature implemented?

Reading your comment, I am still not sure I understand 100% what you are trying to accomplish. Are you aware that you can create your own task by subclassing `Task` and for example supply a custom `CompletionHandler`?

Yes, feature-requests are very welcome :). There is no `RecurringTask` class with task_data because it has not been needed/requested yet. And you can create your own task-class to achieve it. Just have in mind that the task-data can not be changed, and will need to be set in the tasks implementation of `OnStartup`. I will add it to the feature-list though :)

**Comment by ruifscouto** [comment]
So, let me try to explain it more clearly:
I have a set of recurring tasks and a set of one-time tasks. Within the one time tasks I have tasks that will run once without any data and a task that will run every 5 seconds with changing data (the updated data from one execution should be available to the next execution and so on).

For the recurrent tasks and one-time tasks, without data, one scheduler is fine. For the second case, I need to call the .schedule() method of the scheduler with the time of next execution of the task and the **updated data** for the following execution. Hence, I can't include this task on the scheduler construction, therefore, cannot reduce the usage to a single scheduler.
I've attempted the completionHandler approach but at that point I have to access to the task_data via the ExecutionComplete object and would need to be able to specify that task_data for the next execution

I think the keyword that was missing is that I need task_data to be updated in between executions, so I think I can only use the .schedule() method of a scheduler for that.

So the feature request would be to have a way to update the task_data in between executions - possibly by enriching the Execution object with task_data? Not quite sure on this point.

Regarding time-frames I'm not sure how busy you are or if you have other priorities. For now I can solve the logging by suppressing logs from that package, which is not ideal, and something I wouldn't like to keep long.

**Comment by kagkarlsson** [comment]
Ok, now I think I have the full picture. It will look into rescheduling tasks with new task-data. Should be a pretty small change.

Thanks alot for the feature-feedback!

**Comment by kagkarlsson** [comment]
A small detail, when you reschedule with new data, how do you know what the new data will be? Is it runtime knowledge or static (like i = i + 1)?

**Comment by ruifscouto** [comment]
The new data is a set of indexes that is incremented to my desire in each execution. I pass the updated object to a new instance object which is then used with the schedule method of the scheduler.

**Comment by kagkarlsson** [comment]
Struggling a bit with finding the time to add these features, so it might take some days..

**Comment by ruifscouto** [comment]
No problem. Thanks for the heads up.

**Comment by kagkarlsson** [comment]
Version `2.3`, which exposes a `SchedulerClient` on the `ExecutionContext` in the `ExecutionHandler` interface, is soon available in maven central. Should let you get rid of the double schedulers for now.

**Comment by ruifscouto** [comment]
Hey, the latest version does allow me to get rid of the double schedulers. Thank you.
So should I close this and create a feature request for recurrent tasks with task_data? or create a new issue for that?

**Comment by kagkarlsson** [comment]
Great! Then I will close this one. 
Could you register a new issue, preferably feature-request, for recurring tasks with variable task-data?

**Comment by kagkarlsson** [comment]
🎉 This issue has been resolved in `v15.2.0` ([Release Notes](https://github.com/kagkarlsson/db-scheduler/releases/tag/v15.2.0))
